Zend/zend_compile.c

```diff
@@ -4516,22 +4517,93 @@ static zend_result zend_compile_func_array_slice(znode *result, zend_ast_list *a
 }
 /* }}} */
 
-static zend_result zend_try_compile_special_func(znode *result, zend_string *lcname, zend_ast_list *args, zend_function *fbc, uint32_t type) /* {{{ */
+static int find_frameless_function_offset(int arity, void *handler)
 {
-	if (CG(compiler_options) & ZEND_COMPILE_NO_BUILTINS) {
+	void **handlers = zend_flf_handlers;
+	void **current = handlers;
+	while (current) {
+		if (*current == handler) {
+			return current - handlers;
+		}
+		current++;
+	}
+
+	return (uint32_t)-1;
+}
+
+static zend_result zend_compile_frameless_icall(znode *result, zend_ast_list *args, zend_function *fbc, uint32_t type)
+{
+	if (ZEND_OBSERVER_ENABLED) {
 		return FAILURE;
 	}
 
-	if (fbc->type != ZEND_INTERNAL_FUNCTION) {
-		/* If the function is part of disabled_functions, it may be redeclared as a userland
-		 * function with a different implementation. Don't use the VM builtin in that case. */
+	if (type != BP_VAR_R) {
 		return FAILURE;
 	}
 
-	if (zend_args_contain_unpack_or_named(args)) {
+	if (ZEND_USER_CODE(fbc->type)) {
+		return FAILURE;
+	}
+
+	const zend_frameless_function_info *frameless_function_info = fbc->internal_function.frameless_function_infos;
+	if (!frameless_function_info) {
+		return FAILURE;
+	}
+
+	if (args->children > 3) {
 		return FAILURE;
 	}
 
+	/* We don't support named arguments or argument unpacking. */
+	for (uint32_t i = 0; i < args->children; i++) {
```

Comment:
Can use `if (zend_args_contain_unpack_or_named(args)) {` here instead in principle

Reply:
Actually I don't think you need this because `zend_compile_frameless_icall` is only called from `zend_try_compile_special_func` which already does that check?

Reply:
Indeed, good catch!